### PR TITLE
refactor: Refactor width persistence

### PR DESCRIPTION
**Pull request by STATICHIT**
English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Background and solution
refactor: Refactor width persistence

### What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
  - Enhanced plugin width management using local storage for personalized settings.
  - Introduced a new `addons` prop in multiple components, improving configuration and functionality.

- **Bug Fixes**
  - Ensured components correctly reference the updated `addons` structure for improved reliability.

- **Documentation**
  - Updated component exports to include the new `addons` prop, aiding developers in implementation.

- **Chores**
  - Standardized handling of width and option properties across various plugins for consistency.

### CODE REVIEW

**Review by lizhijie429 — APPROVED** [review]

File: packages/controller/src/useLayout.js [hunk 1]
```diff
@@ -176,6 +151,26 @@ const getPluginState = () => layoutState.plugins
 const isEmptyPage = () => layoutState.pageStatus?.state === PAGE_STATUS.Empty
 
 export default () => {
+  let theLocalStorage = '{}'
+  const myLocalStorage = localStorage.getItem('plugin')
+  if (typeof myLocalStorage === 'string' && myLocalStorage.toString !== 'undefined') {
```
Comment (hexqi) [inline packages/controller/src/useLayout.js]: 前面条件为真的话，myLocalStorage是字符串，myLocalStorage.toString肯定是一个函数，这里判断undefined是为什么？

File: packages/controller/src/useLayout.js [hunk 2]
```diff
@@ -176,6 +151,26 @@ const getPluginState = () => layoutState.plugins
 const isEmptyPage = () => layoutState.pageStatus?.state === PAGE_STATUS.Empty
 
 export default () => {
+  let theLocalStorage = '{}'
+  const myLocalStorage = localStorage.getItem('plugin')
```
Comment (hexqi) [inline packages/controller/src/useLayout.js]: 注意下变量命名，theLocalStorage、myLocalStorage都是非常奇怪的名字，要起能看出变量含义的名字

File: packages/controller/src/useLayout.js [hunk 3]
```diff
@@ -176,6 +151,26 @@ const getPluginState = () => layoutState.plugins
...
+  if (typeof myLocalStorage === 'string' && myLocalStorage.toString !== 'undefined') {
+    theLocalStorage = myLocalStorage
+  }
+  theLocalStorage = JSON.parse(theLocalStorage)
```
Comment (hexqi) [inline packages/controller/src/useLayout.js]: JSON.parse最好加上try catch，可能会报错

File: packages/controller/src/useLayout.js [hunk 4]
```diff
@@ -176,6 +151,26 @@ const getPluginState = () => layoutState.plugins
...
+
+  const pluginWidthStorage = useStorage('plugin', theLocalStorage)
+
+  const getPluginWidth = (name) => {
```
Comment (hexqi) [inline packages/controller/src/useLayout.js]: 可以简化成：
`const getPluginWidth = (name) => pluginWidthStorage.value[name]?.width || 300`

File: packages/design-core/src/App.vue [hunk 5]
```diff
@@ -5,7 +5,7 @@
       <design-toolbars></design-toolbars>
       <div class="tiny-engine-main">
         <div class="tiny-engine-left-wrap">
-          <design-plugins :render-panel="plugins.render" @click="toggleNav"></design-plugins>
+          <design-plugins :render-panel="plugins.render" :Addons="addons" @click="toggleNav"></design-plugins>
```
Comment (hexqi) [inline packages/design-core/src/App.vue]: 需要与其他属性保持一样的风格，`:Addons` ==> `:addons`

File: packages/design-core/src/App.vue [hunk 6]
```diff
@@ -46,7 +47,7 @@ import blockPlugin from '@opentiny/tiny-engine-plugin-block'
 import materials from '@opentiny/tiny-engine-plugin-materials'
 import { useBroadcastChannel } from '@vueuse/core'
 import { constants } from '@opentiny/tiny-engine-utils'
-
+import { useStorage } from '@vueuse/core'
```
Comment (hexqi) [inline packages/design-core/src/App.vue]: import和后面的语句之间空一行

File: packages/design-core/src/App.vue [hunk 7]
```diff
@@ -76,7 +77,14 @@ export default {
...
 
-    const { layoutState } = useLayout()
+    const plugin = {}
+    addons.plugins.forEach((item) => {
+      plugin[item.id] = {}
+      plugin[item.id].width = item.options?.width || 300
+    })
+    useStorage('plugin', plugin)
+
+    const { layoutState } = useLayout() //执行useLayout
```
Comment (hexqi) [inline packages/design-core/src/App.vue]: 这个注释没有什么实际的意义，可以删掉

File: packages/design-core/src/App.vue [hunk 8]
```diff
@@ -76,7 +77,14 @@ export default {
       jsClose: null
     })
 
-    const { layoutState } = useLayout()
+    const plugin = {}
+    addons.plugins.forEach((item) => {
+      plugin[item.id] = {}
+      plugin[item.id].width = item.options?.width || 300
```
Comment (hexqi) [inline packages/design-core/src/App.vue]: 这两行可以合成一条语句：
```js
plugin[item.id] = {
  width: item.options?.width || 300
}
```

File: packages/controller/src/useLayout.js (hunk 1 above)
Comment (STATICHIT) [reply packages/controller/src/useLayout.js]: 因为使用vueuse的useStorage时，如果存储时传入的是空，实际存入的会是undefined字符串。这里需要特别判断，否则后续使用的将是undefined字符串。这一行就是对存储内容是否为“undefined"的判断。开发过程已和导师一起验证过。

Comment (hexqi) [reply packages/controller/src/useLayout.js]: 那应该直接判断 myLocalStorage !== 'undefined'

Comment (STATICHIT) [reply packages/controller/src/useLayout.js]: 好的

**Review by hexqi — APPROVED** [review]

**Pull request merged**